LibC+LibELF: Share heap blocks between LibC and DynamicLoader

gunnarbeutner opened the pull request:
Previously LibC and DynamicLoader had their own separate heaps because DynamicLoader has its own statically linked private copy of LibC. This change updates LibC to forward calls to the allocation functions to the DynamicLoader.

This saves about 448kB of committed pages per process and about 28kB of dirty pages.

timschumi reviewed (commented):
One bullet point on the ever-lasting TODO list is to move more parts of the initialization into the `crt*` files so that we can support static binaries a bit better (something about not having the `DynamicLoader`). I feel like this change might throw a wrench into those plans in particular. (In fact, it seems to me like this change would leave us without a malloc implementation to delegate to?)

Also, please note the linting failures, which currently keep tests from running.

timschumi commented:
More background on the initialization discussion can be found in #15373.

gunnarbeutner commented:
Aye, that makes sense. I'll see what I can do about that. :)

gunnarbeutner commented on Userland/Libraries/LibC/bits/malloc_integration.h:
```diff
@@ -0,0 +1,21 @@
+typedef void* (*ReallocFunction)(void*, size_t);
+typedef void (*SerenityDumpMallocStatsFunction)();
+typedef bool (*SerenityHeapIsStableFunction)();
+typedef bool (*SerenitySetAllocationEnabledFunction)(bool);
```
That's a lot better. Fixed. :+1:

gunnarbeutner commented:
Here's what I changed:

* Updated static LibC so that it contains the allocation functionality in a way that doesn't require DynamicLoader/crt0 to patch anything.

Also, bonus change:

**LibC: Increase number of usable chunks for a malloc size class**

This decreases one of the the size classes from 32752 bytes to 32736 bytes because previously there was only one usable chunk (on x86_64):

`chunks = (65536 (block size) - 60 (header on x86_64)) / 32752 = 1.99`

This does waste a few more bytes on i686 but I think it's a fair trade-off without making the size classes more complicated, e.g. by
picking different ones based on the architecture.

timschumi commented:
Looks like it's now using the `errno` from `DynamicLoader`, which is likely not what we want (because at that point we are also getting into multithreading-related things).

gunnarbeutner commented:
You're right. :)

I've split the patch into a few more commits and fixed the errno issue. Unfortunately we're back to good old C style typedefs now though.

timschumi commented on Userland/Libraries/LibC/malloc.cpp:
```diff
@@ -537,17 +538,15 @@ static void free_impl(void* ptr)
 
-    if (ptr_or_error.is_error()) {
-        errno = ptr_or_error.error().code();
-        return nullptr;
-    }
+    if (ptr_or_error.is_error())
+        return ptr_or_error.release_error();
```
Does `TRY` not work here?

```
auto ptr = TRY(malloc_impl(size, 16, CallerWillInitializeMemory::No));`
```

timschumi commented on Userland/Libraries/LibC/malloc.cpp:
```diff
@@ -566,27 +565,23 @@ void free(void* ptr)
 
-    if (ptr_or_error.is_error()) {
-        errno = ptr_or_error.error().code();
-        return nullptr;
-    }
+    if (ptr_or_error.is_error())
+        return ptr_or_error.release_error();
```
Same here.

timschumi commented on Userland/Libraries/LibC/malloc.cpp:
```diff
@@ -626,33 +614,36 @@ size_t malloc_size(void const* ptr)
         return ptr;
     }
-    auto* new_ptr = malloc(size);
+    auto new_ptr_or_error = __malloc(size);
+    if (new_ptr_or_error.is_error())
+        return new_ptr_or_error;
+    auto* new_ptr = new_ptr_or_error.value();
```
And here.

timschumi commented on Userland/Libraries/LibELF/DynamicLinker.cpp:
```diff
@@ -268,33 +278,29 @@ static void initialize_libc(DynamicObject& libc)
+    update_libc_variable("__dlsym"sv, __dlsym);
+    update_libc_variable("__dladdr"sv, __dladdr);
+
+    update_libc_variable("__malloc"sv, __malloc);
+    update_libc_variable("__free"sv, __free);
+    update_libc_variable("__calloc"sv, __calloc);
+    update_libc_variable("__posix_memalign"sv, __posix_memalign);
+    update_libc_variable("__aligned_alloc"sv, __aligned_alloc);
+    update_libc_variable("__malloc_size"sv, __malloc_size);
+    update_libc_variable("__malloc_good_size"sv, __malloc_good_size);
+    update_libc_variable("__realloc"sv, __realloc);
+    update_libc_variable("__serenity_dump_malloc_stats"sv, __serenity_dump_malloc_stats);
+    update_libc_variable("___heap_is_stable"sv, ___heap_is_stable);
+    update_libc_variable("___set_allocation_enabled"sv, ___set_allocation_enabled);
```
This could probably use a comment in the sense of "This delegates memory management to the copy of the implementation in the DynamicLoader, which keeps us from allocating a second memory pool".
